File: sdk/runanywhere-swift/Sources/RunAnywhere/Public/Extensions/RunAnywhere+SpeakerDiarization.swift
//
//  RunAnywhere+SpeakerDiarization.swift
//  RunAnywhere SDK
//
//  Public API for Speaker Diarization (B12, §8) — identifies which speaker
//  produced each chunk of an audio stream.
//
//  Status: TODO-stub facade. The C ABI exists in runanywhere-commons
//    (rac_speaker_diarization_init / _process / _destroy — see
//     `sdk/runanywhere-commons/include/rac/features/speaker_diarization/
//      rac_speaker_diarization.h`)
//  but:
//    1. `rac_speaker_diarization.cpp` is a stub returning
//       `RAC_ERROR_FEATURE_NOT_AVAILABLE`.
//    2. The header is NOT yet exposed through the Swift `CRACommons`
//       umbrella header (`Sources/RunAnywhere/CRACommons/include/CRACommons.h`),
//       so the symbols cannot be called from Swift until the header is added.
//
//  This facade therefore:
//    - `loadDiarizationModel(_:)` throws `SDKException.runtime(.featureNotAvailable, …)`.
//    - `diarize(audio:)` logs a warning via `os_log` and returns `[]`.
//    - `unloadDiarization()` is a no-op.
//
//  TODO(diarization): when the real C++ implementation lands, also
//  update `CRACommons/include/CRACommons.h` to `#include "rac_speaker_diarization.h"`
//  and ship the header. Then replace the bodies below with direct C ABI
//  calls (see the archived richer implementation in git history for
//  `DiarizationSession` scaffolding). The public surface below stays the
//  same.
//

import Foundation
import os

// MARK: - Public Types

/// One speaker segment returned by `RunAnywhere.diarize(audio:)`.
///
/// Represents a contiguous span of audio attributed to a single speaker.
public struct SpeakerSegment: Sendable, Hashable, Codable {

    /// Zero-based speaker index. Stable within a single diarization session
    /// (consecutive calls on the same handle), not across sessions.
    public let speaker: Int

    /// Segment start time in milliseconds from the beginning of the audio.
    public let startMs: Int

    /// Segment end time in milliseconds from the beginning of the audio.
    public let endMs: Int

    public init(speaker: Int, startMs: Int, endMs: Int) {
        self.speaker = speaker
        self.startMs = startMs
        self.endMs = endMs
    }
}

// MARK: - Public Facade

public extension RunAnywhere {

    /// Load the speaker-diarization model.
    ///
    /// - Parameter modelPath: Filesystem path to the diarization model.
    /// - Throws: `SDKException.runtime(.featureNotAvailable, _)` until
    ///   the native diarization implementation has been integrated
    ///   into commons and exposed through the Swift `CRACommons`
    ///   umbrella header.
    static func loadDiarizationModel(_ modelPath: String) async throws {
        guard isInitialized else {
            throw SDKException.general(.notInitialized, "SDK not initialized")
        }
        os_log("Speaker diarization not yet available in commons. modelPath=%{public}@",
               log: .default, type: .info, modelPath)
        throw SDKException.runtime(
            .featureNotAvailable,
            "Speaker diarization is not yet integrated in runanywhere-commons."
        )
    }

    /// True if a diarization session is loaded. Always `false` while the
    /// feature is stubbed.
    static var isDiarizationLoaded: Bool {
        get async { false }
    }

    /// Run speaker diarization on a buffer of PCM float samples (16 kHz mono).
    ///
    /// - Parameter audio: Raw IEEE-754 single-precision PCM samples as `Data`
    ///   (4 bytes per sample).
    /// - Returns: Array of `SpeakerSegment` ordered by `startMs`. Empty
    ///   array while the native diarization feature is not yet available
    ///   (a warning is logged so the gap is diagnosable).
    static func diarize(audio: Data) async throws -> [SpeakerSegment] {
        guard isInitialized else {
            throw SDKException.general(.notInitialized, "SDK not initialized")
        }
        os_log("diarize: feature not yet available in commons (stub). Returning []. audioBytes=%d",
               log: .default, type: .info, audio.count)
        return []
    }

    /// Release the diarization session and free its resources.
    ///
    /// No-op while the feature is stubbed.
    static func unloadDiarization() async {
        // No resources to release while stubbed.
    }
}

File: sdk/runanywhere-kotlin/src/commonMain/kotlin/com/runanywhere/sdk/public/extensions/RunAnywhere+SpeakerDiarization.kt
/*
 * Copyright 2026 RunAnywhere SDK
 * SPDX-License-Identifier: Apache-2.0
 *
 * Public API for Speaker Diarization (B12, §8) — identifies which speaker
 * produced each chunk of an audio stream.
 *
 * Mirrors Swift `RunAnywhere+SpeakerDiarization.swift`. Platform actuals
 * delegate to the C ABI in runanywhere-commons:
 *   • rac_speaker_diarization_init(model_path, out_handle)
 *   • rac_speaker_diarization_process(handle, samples, count, out_json)
 *   • rac_speaker_diarization_destroy(handle)
 *
 * The native implementation is currently a stub that returns
 * RAC_ERROR_FEATURE_NOT_AVAILABLE; actuals log a warning and return an
 * empty segment list so callers get a well-defined "feature not ready"
 * signal rather than a crash.
 */

package com.runanywhere.sdk.public.extensions

import com.runanywhere.sdk.public.RunAnywhere

/**
 * One speaker segment returned by [RunAnywhere.diarize].
 *
 * Represents a contiguous span of audio attributed to a single speaker.
 *
 * @property speaker Zero-based speaker index. Stable within one session.
 * @property startMs Segment start time in milliseconds.
 * @property endMs   Segment end time in milliseconds.
 */
data class SpeakerSegment(
    val speaker: Int,
    val startMs: Long,
    val endMs: Long,
)

/**
 * Load the speaker-diarization model.
 *
 * Optional — [diarize] also lazily loads if no model is active. Throws
 * [com.runanywhere.sdk.foundation.errors.SDKException] if the feature
 * is not yet available in commons.
 *
 * @param modelPath Filesystem path to the diarization model.
 */
expect suspend fun RunAnywhere.loadDiarizationModel(modelPath: String)

/**
 * True if a diarization session is currently loaded.
 */
expect val RunAnywhere.isDiarizationLoaded: Boolean

/**
 * Run speaker diarization on a buffer of PCM audio.
 *
 * @param audio IEEE-754 single-precision PCM samples (little-endian, 4
 *   bytes per sample, 16 kHz mono).
 * @return Segments ordered by [SpeakerSegment.startMs]. Empty when either
 *   no speech was detected or the native feature is not yet available
 *   (a warning is logged in the latter case).
 */
expect suspend fun RunAnywhere.diarize(audio: ByteArray): List<SpeakerSegment>

/**
 * Release the diarization session and its native resources.
 */
expect suspend fun RunAnywhere.unloadDiarization()
